File: android/app/src/main/java/app/opensleep/domain/sonar/AwakeWhenHighActivity.kt
package app.opensleep.domain.sonar

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import androidx.core.content.ContextCompat

class AwakeWhenHighActivity(private val context: Context) {
    private val classifiers = mutableMapOf<String, AverageActivityOverThreshold>()
    private var lastAwakeDetected = 0L

    private val receiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent?) {
            if (intent == null) return
            val sensorType = intent.getStringExtra("action_raw_activity_sensor") ?: return
            val value = intent.getFloatExtra("action_raw_activity_data", 0f)

            // Ignore non-accelerometer/watch sensors (such as SONAR) to prevent noise from triggering false awake intervals
            if (sensorType != "PHONE_ACCEL" && sensorType != "SMARTWATCH" && sensorType != "BLE_ACCEL") {
                return
            }

            val classifier = classifiers.getOrPut(sensorType) {
                val threshold = if (sensorType == "SMARTWATCH" || sensorType == "BLE_ACCEL") {
                    2.5f
                } else {
                    1.5f
                }
                AverageActivityOverThreshold(threshold)
            }

            if (classifier.update(value)) {
                lastAwakeDetected = System.currentTimeMillis()
            }
        }
    }

    init {
        ContextCompat.registerReceiver(
            context,
            receiver,
            IntentFilter("action_raw_activity"),
            ContextCompat.RECEIVER_NOT_EXPORTED
        )
    }

    fun isAwake(): Boolean {
        return System.currentTimeMillis() - lastAwakeDetected < 60000
    }

    fun stop() {
        context.unregisterReceiver(receiver)
        classifiers.clear()
        lastAwakeDetected = 0L
    }
}

File: ios/opensleep/opensleep/Services/Sonar/AwakeWhenHighActivity.swift
import Foundation

class AwakeWhenHighActivity {
    private var classifiers: [String: AverageActivityOverThreshold] = [:]
    private var lastAwakeDetected: Date = Date.distantPast
    private let lock = NSLock()
    private var notificationObserver: Any?

    init() {
        notificationObserver = NotificationCenter.default.addObserver(
            forName: NSNotification.Name("action_raw_activity"),
            object: nil,
            queue: nil
        ) { [weak self] notification in
            self?.handleRawActivity(notification)
        }
    }

    private func handleRawActivity(_ notification: Notification) {
        lock.lock(); defer { lock.unlock() }
        guard let userInfo = notification.userInfo,
              let sensorType = userInfo["sensor"] as? String,
              let value = userInfo["data"] as? Float else { return }

        // Ignore non-accelerometer/watch sensors (such as SONAR) to prevent noise from triggering false awake intervals
        guard sensorType == "PHONE_ACCEL" || sensorType == "SMARTWATCH" || sensorType == "BLE_ACCEL" else {
            return
        }

        let classifier = classifiers[sensorType] ?? {
            let threshold: Float = (sensorType == "SMARTWATCH" || sensorType == "BLE_ACCEL") ? 2.5 : 1.5
            let newClassifier = AverageActivityOverThreshold(threshold: threshold)
            classifiers[sensorType] = newClassifier
            return newClassifier
        }()

        if classifier.update(activity: value) {
            lastAwakeDetected = Date()
        }
    }

    func isAwake() -> Bool {
        lock.lock(); defer { lock.unlock() }
        return Date().timeIntervalSince(lastAwakeDetected) < 60.0
    }

    func stop() {
        lock.lock(); defer { lock.unlock() }
        if let observer = notificationObserver {
            NotificationCenter.default.removeObserver(observer)
            notificationObserver = nil
        }
        classifiers.removeAll()
        lastAwakeDetected = Date.distantPast
    }

    deinit {
        if let observer = notificationObserver {
            NotificationCenter.default.removeObserver(observer)
        }
    }
}
